packages/concerto-core/lib/serializer/jsongenerator.js

```diff
@@ -142,18 +144,40 @@ class JSONGenerator {
             }
         }
 
+        // by default we set the $class on the result
         result.$class = classDeclaration.getFullyQualifiedName();
+
+        // if we are in the inferClass mode and this is not the root instance
+        // we attempt to remove the $class
+        if(this.inferClass && !parameters.isRoot && parameters.field) {
+            const fieldClassDecl = parameters.modelManager.getType(parameters.field.getFullyQualifiedTypeName());
+            // if the $class cannot be unambigiously inferred from the type of the field in the model
+            // we have to include it, but we attempt to shorten it, if the object and the field are in the same ns
+            if(fieldClassDecl.getAssignableClassDeclarations().length > 2) {
```

Comment:
It looks like getAssignableClassDeclarations finds all equal or derived types in all known models. If so, then this looks like it keeps the $class when there exist at least three (why not two?) matching types, but that seems like an unreliable property because a new model can extend a concept at any time, leading to the case later on where something that "was unambigious is now ambigious".

I'd recommend omitting the class only when it's exactly equal to the declared field type. That is a stable property that can't break with the addition of new models.

Reply:
> I'd recommend omitting the class only when it's exactly equal to the declared field type. That is a stable property that can't break with the addition of new models.

I like this idea — it limits the scope of the optimisation, but should still give us some value for simple cases.